src/elements/rating/rating.js

```diff
@@ -13,75 +13,133 @@
     })
     .directive('smRating', smRating);
 
-  /*
-   * TODO(mxth): do not rely on ngRepeat
-   * TODO(mxth): do not rely on ngClass
-   */
-  var template =
-    '<div ng-mouseleave="ctrl.reset()" ng-keydown="ctrl.onKeydown($event)" tabindex="0" role="slider"' +
-      'aria-valuemin="0" aria-valuemax="{{ctrl.range.length}}" aria-valuenow="{{ctrl.value}}"' +
-      'class="ui rating" ng-class="ctrl.hoverValue > -1 && (ctrl.stateHoverParent || \'selected\')">' +
-      '<i ng-repeat="r in ctrl.range track by $index" ng-mouseenter="ctrl.enter($index + 1)" ng-click="ctrl.rate($index + 1)"' +
-        'class="icon" ng-class="[' +
-          '($index < ctrl.value && (r.stateActive || \'active\')) || \'\',' +
-          '($index < ctrl.hoverValue && (r.stateHover || \'selected\')) || \'\']">' +
-      '</i>' +
-    '</div>';
-
-  function smRating() {
+  smRating.$inject = ['ratingConfig'];
+
+  function smRating(ratingConfig) {
     return {
       restrict:'EA',
-      replace: true,
       require: ['smRating', 'ngModel'],
-      scope: {
-        readonly: '=?',
-        onHover: '&',
-        onLeave: '&'
-      },
       controller: smRatingController,
-      controllerAs: 'ctrl',
-      bindToController: true,
-      template: template,
-      link: function(scope, element, attrs, ctrls) {
-        var ratingCtrl = ctrls[0], ngModelCtrl = ctrls[1];
-        ratingCtrl.init(ngModelCtrl, element);
-      }
+      compile: compile
     };
+
+    function compile(element, attrs) {
+      element.addClass('ui rating');
+      element.attr('tabindex', 0);
+      element.attr('role', 'slider');
+      element.attr('aria-valuemin', 0);
+
+      element.addClass(angular.isDefined(attrs.type) ? attrs.type : ratingConfig.type);
+      element.addClass(angular.isDefined(attrs.size) ? attrs.size : ratingConfig.size);
+
+      return postLink;
+    }
+
+    function postLink(scope, element, attr, ctrls) {
+      var ratingCtrl = ctrls[0], ngModelCtrl = ctrls[1];
+      ratingCtrl.init(ngModelCtrl);
+    }
   }
 
-  smRatingController.$inject = ['$scope', '$attrs', 'ratingConfig'];
+  smRatingController.$inject = ['$scope', '$element', '$attrs', 'ratingConfig', '$parse'];
+
+  function smRatingController($scope, $element, $attrs, ratingConfig, $parse) {
+    var self = this;
+
+    self.$scope = $scope;
+    self.$element = $element;
+    self.$attrs = $attrs;
+    self.ratingConfig = ratingConfig;
+
+    self.ngModelCtrl  = { $setViewValue: angular.noop };
```

Comment:
I borrowed this from ui-bootstrap. I think the purpose is to let `rate` function be called before `link` without throwing `undefined` error